Dask on demand post (#10)
ian-whitestone/ian-whitestone.github.io: _posts/2020-12-21-dask-on-demand.md

    "serviceAccounts": [
        {
            "email": f"{project_number}-compute@developer.gserviceaccount.com",
            "scopes": [
                "https://www.googleapis.com/auth/devstorage.read_only",
                "https://www.googleapis.com/auth/logging.write",
                "https://www.googleapis.com/auth/monitoring.write",
                "https://www.googleapis.com/auth/servicecontrol",
                "https://www.googleapis.com/auth/service.management.readonly",
                "https://www.googleapis.com/auth/trace.append",
            ],
        }
    ],
}
```

Once we have this instance config, the hard work is done. The API call to create the instance can be done in a few lines:

```python
from googleapiclient import discovery

compute = discovery.build("compute", "v1", cache_discovery=False)
operation = (
    compute.instances()
    .insert(project=project_id, zone=cluster_zone, body=instance_config)
    .execute()
)
```

The command above returns an object containing metadata about the operation you just initiated. You can use this to repeatedly check the status of the operation until it is completed.

## Wrapping everything into a class

> With the core fundamentals outlined above, I'll go through the high level details of a Python class I created to perform all this functionality. The class will also contain some nice extras we need to perform the capabilities I outlined in the beginning of this post, like instantiating a new Dask client or tearing down the GCE instance. If you want to skip the explanation, you can just see the full source code [here](https://gist.github.com/ian-whitestone/42fc0d142cc7f63c79f1e693d58a13a9).

Author: ```suggestion
With the core fundamentals outlined above, I'll go through the high level details of a Python class I created to perform all this functionality. The class will also contain some extra functionality we need to satisfy the capabilities I outlined in the beginning of this post, like instantiating a new Dask client or tearing down the GCE instance. If you want to skip the explanation, you can just see the full source code [here](https://gist.github.com/ian-whitestone/42fc0d142cc7f63c79f1e693d58a13a9).
```

    gcp_cluster_zone: str
    gcp_instance_name: str
    gcp_docker_image: str

    machine_type: str
    num_workers: int
    threads_per_worker: int
    memory_per_worker_gb: int

    def __post_init__(self):
        self._validate_machine_type()
        self.compute = discovery.build("compute", "v1", cache_discovery=False)
        self.disk_image_name, self.disk_image_link = self._get_latest_image()
        self.create()
        self.cluster_host_ip_address = self._get_cluster_ip_address()
        self._wait_until_cluster_is_ready()
        self.client = self.create_client()
```

With this setup, you can create a cluster by running `cluster = Cluster(gcp_project_id="my_project_id", ...)`. Alternatively, you can put in some defaults directly in the class so you don't have to provide all the arguments each time. When the `Cluster` class is instantiated, the following will happen:

1. We'll run some checks to make sure the `machine_type` provided by the user is valid (i.e. check if it's named properly, or if the number of workers specified makes sense given the number of cores available on the machine).
2. Create a new compute object we'll use to interact with the GCE APIs.

> 3. Programatically grab the latest COS image to create our GCE instance with, so we're hardcoding everything like I did above for `source_image` and `disk_image_name` which are required by the instance config object.

Author: ```suggestion
3. Programatically grab the latest COS image to create our GCE instance with, so we're not hardcoding everything like I did above for `source_image` and `disk_image_name`, which are required by the instance config object.
```

> 4. Create the GCE instance. Our class will also have an `instance_config` property which is required by the API call.

Author: ```suggestion
4. Create the GCE instance.
```

                    memory_per_worker_gb=memory_per_worker_gb,
                    machine_type=machine_type,
                    gcp_instance_name=gcp_instance_name,
                    gcp_cluster_zone=gcp_cluster_zone,
                )

                # update the args/kwargs with the newly created client
                # see source code for how to handle updating the args
                # https://gist.github.com/ian-whitestone/42fc0d142cc7f63c79f1e693d58a13a9
                kwargs["client"] = cluster.client
                return func(*new_args, **kwargs)
            finally:
                if cluster is not None:
                    cluster.teardown()
        return wrapper

    return decorator
```

The decorator runs some checks to see if a user has already provided a Dask client to their function, and if so, just runs the function as is without creating a new cluster. If a client has not been provided, the decorator spins up a Dask cluster with a connected client, and passes that client to the function for its use. Regardless of whether the function runs successfully or not, the cluster is automatically torn down.

> ## Examples workflows

Author: ```suggestion
## Example workflows
```
